Add additional logging for notification conditions

sghill opened the pull request:
- Extracted Condition interface that let's us log both if the condition
  was met and if the user preference allowed the notification to be sent
- Conditions that have interesting logic are unit-tested
- Moves to the message supplier style of logging, which is not invoked
  unless that level of logging is enabled
- Introduces a build key format - [ProjectFullDisplayName #<build>] to
  better indicate which build the messaging is coming from

Fixes #508

timja commented:
Looks good, I'll try it out soon just to make sure it's fine

timja commented on src/main/java/jenkins/plugins/slack/decisions/OnBackToNormal.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public boolean isMetBy(Context context) {
+        Result previousResult = context.previous.getResult();
```
This throws a nullpointerexception if there's no previous builds

timja commented:
Here's an example log from running a build with all options ticked:
```
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.SlackNotifier prebuild
INFO: Performing start notifications
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.ActiveNotifier started
INFO: [freestyle 1 #4] was not caused by SCM Trigger
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.StandardSlackService publish
INFO: Posting succeeded
Feb 14, 2019 8:32:27 AM hudson.model.Run execute
INFO: freestyle 1 #4 main build action completed: SUCCESS
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.SlackNotifier perform
INFO: Performing complete notifications
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.ActiveNotifier completed
INFO: [freestyle 1 #4] found #3 as previous completed, non-aborted build
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.decisions.Condition test
INFO: [freestyle 1 #4] will send OnSuccessNotification because build matches and user preferences allow it
Feb 14, 2019 8:32:27 AM jenkins.plugins.slack.ActiveNotifier$MessageBuilder getStatusMessage
INFO: [freestyle 1 #4] selected #3 as previous completed, non-aborted build
Feb 14, 2019 8:32:28 AM jenkins.plugins.slack.StandardSlackService publish
INFO: Posting succeeded
Feb 14, 2019 8:32:28 AM jenkins.plugins.slack.SlackNotifier perform
INFO: Performing finalize notifications
```

timja commented:
@sghill What do you think about moving some of those messages to the build log instead of the system log?

I don't like how in freestyle projects its silent and you can't even tell its done anything

timja commented:
(Code is good except for the missing null checks around previous build)

sghill commented:
Thanks for the quick review @timja!

I updated the `Context` to never have a nullable previous result (falls back to success, just like before). Tests have been added for this as well. 

> I don't like how in freestyle projects its silent and you can't even tell its done anything

I agree, that's a great idea.

> What do you think about moving some of those messages to the build log instead of the system log?

I introduced a `SystemAndUserLogger` which will write info-level logs to both the build log and the system log. I'd prefer them in both places because it can be helpful to only have to look at the system log as an operator. This is also covered by tests.

Here's an example of logs in the build (they include the plugin name so we know where they came from):

```
[Slack Notifications] found #5 as previous completed, non-aborted build
[Slack Notifications] will send OnSuccessNotification because build matches and user preferences allow it
Finished: SUCCESS
```

I moved the lifecycle messages to the FINE level. Here's an example of the system log at INFO level:

```
[my-freestyle-job #6] found #5 as previous completed, non-aborted build
[my-freestyle-job #6] will send OnSuccessNotification because build matches and user preferences allow it
```

and at FINE level:

```
[my-freestyle-job #6] Performing start notifications
[my-freestyle-job #6] was not caused by SCM Trigger
[my-freestyle-job #6] Performing complete notifications
[my-freestyle-job #6] found #5 as previous completed, non-aborted build
[my-freestyle-job #6] does not match OnAbortedNotification condition
[my-freestyle-job #6] does not match OnSingleFailureNotification condition
[my-freestyle-job #6] does not match OnRepeatedFailureNotification condition
[my-freestyle-job #6] does not match OnNotBuiltNotification condition
[my-freestyle-job #6] does not match OnBackToNormalNotification condition
[my-freestyle-job #6] will send OnSuccessNotification because build matches and user preferences [my-freestyle-job #6] Performing finalize notifications
```

timja commented on src/main/java/jenkins/plugins/slack/ActiveNotifier.java:
```diff
@@ -1,71 +1,57 @@
-import java.util.List;
-import java.util.Set;
-import java.util.logging.Logger;
+import java.util.*;
```
can you revert these * import changes please?

sghill replied:
Done

timja commented on src/main/java/jenkins/plugins/slack/ActiveNotifier.java:
```diff
@@ -1,71 +1,57 @@
+    private final TokenExpander tokenExpander;
 
-    public ActiveNotifier(SlackNotifier notifier, BuildListener listener) {
+    public ActiveNotifier(SlackNotifier notifier, Function<AbstractBuild<?, ?>, SlackService> slackFactory, BuildAwareLogger log, TokenExpander tokenExpander) {
```
I've searched github and can't see any other users of this constructor so should be ok

timja commented on src/main/java/jenkins/plugins/slack/decisions/OnRepeatedFailure.java:
```diff
@@ -0,0 +1,31 @@
+import jenkins.plugins.slack.SlackNotifier;
+import jenkins.plugins.slack.logging.BuildAwareLogger;
+
+public class OnRepeatedFailure implements Condition {
```
Gah, there should be an OnEveryFailure one imo, this behaviour is just confusing
At least I can now tell what's going on with this code

sghill replied:
Captured this as https://github.com/jenkinsci/slack-plugin/issues/520

timja commented on src/main/java/jenkins/plugins/slack/decisions/Condition.java:
```diff
@@ -0,0 +1,28 @@
+        boolean preferences = userPreferenceMatches();
+        if (isMet) {
+            if (preferences) {
+                log().info(context.currentKey(), "will send " + getClass().getSimpleName() + "Notification because build matches and user preferences allow it");
```
For some reason the logger is coming up as: `UserAndSystemLogger` and not `Condition`

Any idea why, or if its easily fixable to have the correct class name in the logger?

```
Feb 15, 2019 1:24:55 PM jenkins.plugins.slack.logging.UserAndSystemLogger info
INFO: [freestyle 1 #5] will send OnSuccessNotification because build matches and user preferences allow it
```

sghill replied:
It's the way jenkins has defined its log formatter: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/Functions.java#L529

I can update the name to `SlackNotificationsLogger`, which will probably make more sense from the console anyway.

timja commented:
Just the * imports and the question on the logger, otherwise code is good and will get 🚢'ed soon
